relay-sampling/src/lib.rs

```diff
@@ -355,6 +365,125 @@ impl TimeRange {
     }
 }
 
+/// A struct representing the execution context of a decaying function.
+#[derive(Clone, Copy, Debug)]
+struct DecayingFunctionExecutor {
+    decaying_fn: DecayingFunction,
+    external_params: DecayingFunctionExternalParams,
+}
+
+impl DecayingFunctionExecutor {
+    fn execute(&self) -> f64 {
+        match self.decaying_fn {
+            DecayingFunction::Linear { from_sample_rate } => {
+                self.execute_linear_decay(from_sample_rate)
+            }
+            DecayingFunction::Constant => self.external_params.base_sample_rate,
+        }
+    }
+
+    /// Computes the linear decay function at a specific point in time.
+    ///
+    /// The linear decay works in the following way:
+    /// start = 100
+    /// end = 200
+    /// from = 0.7 (70% sample rate)
+    /// base_sample_rate = 0.5 (50% sample rate)
+    /// (from - base_sample_rate) = 0.2 (20% sample rate)
+    ///
+    /// If we have now = 170 it means we are at the 70% progress of the timerange, which means we
+    /// want to inversely reduce the sample rate in the interval (from - base_sample_rate).
+    /// Thus, in the 70% case, we would take only 30% of the (from - base_sample_rate) which
+    /// is equal to 0.06.
+    fn execute_linear_decay(&self, from_sample_rate: f64) -> f64 {
+        let now_timestamp = self.external_params.now.timestamp() as f64;
+        let start_timestamp = self.external_params.start.timestamp() as f64;
+        let end_timestamp = self.external_params.end.timestamp() as f64;
+
+        // We round to 2 digits in order to avoid high-precision sample rates that are more difficult
+        // to work with. Rounding is performed following the nearest integer.
+        let sample_rate_difference =
+            f64::round((from_sample_rate - self.external_params.base_sample_rate) * 100.0) / 100.0;
+        let progress = (now_timestamp - start_timestamp) / (end_timestamp - start_timestamp);
+        let inverse_progress = (1.0 - progress).clamp(0.0, 1.0);
+        self.external_params.base_sample_rate + (sample_rate_difference * inverse_progress)
+    }
+}
+
+/// A struct containing the set of external params required by a decaying function.
+#[derive(Clone, Copy, Debug)]
+struct DecayingFunctionExternalParams {
```

Comment:
I feel that the types involved in this logic are too tightly coupled to the serialization format, in that we differentiate between internal and external params.

In order to decouple them, can we construct something like that from the `SamplingRule`, e.g.

```rust

struct LinearDecay {
    // whatever model parameters a linear function needs. In theory only two (`(k, d)` for `y = kx + d`),
    // but it might make sense to store `(start, end)` timestamps and sample rates as well.
    start_time: DateTime<Utc>,
    end_time: DateTime<Utc>,
    start_rate: f64,
    end_rate: f64,
}

enum DecayFunc {
    // ...
    Linear(LinearDecay),
}

impl SamplingRule {
   fn get_decay_fn(&self) -> DecayFunc {
       // construct DecayFunc from rule here. We can also cache it on the sample rule
   }
}
```

Reply:
Ehm, how would you deserialize this? I wanted to keep my change as small as possible on the existing rule payload.

Reply:
Your proposal makes total sense, indeed, it is the simplest way of doing it but this requires the removal of the timerange, which we can for sure do but we need to be careful with backward compatibility.

Reply:
I would leave the existing rule payload as-is, and create these helper structs independently.